### PR TITLE
Integrate OneLogger across NeMo Speech training

### PR DESCRIPTION
**Pull request by pzelasko**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Integrates opt-in OneLogger lifecycle and dynamic, modality-aware throughput telemetry across NeMo Speech without imposing LLM-specific batch or sequence assumptions.

**Collection**: ASR, TTS, SpeechLM2, Audio, Core/Lightning

# Changelog

- Attach the shared OneLogger callback through `exp_manager` for ModelPT and plain LightningModule workflows when `NEMO_ONE_LOGGER_ENABLED=true`.
- Instrument model construction, eager/deferred dataloader creation, optimizer setup, checkpoint loading/saving, and overall training/validation timing.
- Use reusable context and class decorators so nested operations are paired by identity and lifecycle spans close on exceptions.
- Emit checkpoint success only after synchronous completion or asynchronous finalization; report scheduling failures explicitly.
- Export only after explicit opt-in and only from rank zero.
- Keep OneLogger an optional runtime dependency: disabled mode never imports it, and missing packages degrade to an empty no-op callback registry without affecting training.
- Report the application as `nemo-speech`.
- Move throughput policies into the ASR, TTS, Audio, and SpeechLM2 collection trunks.
- Register policies on semantic mixins, base classes, or concrete models with a shared decorator; normal inheritance carries a policy to compatible subclasses.
- Document the extension contract and a complete future-model example in the shared `speech_throughput` module.
- Report actual rank-local work: ASR audio seconds and target text tokens; TTS input text tokens and output audio seconds; codec, diarization, and audio-model input seconds; DuplexSTT text and audio; speech-to-speech input/output audio at their distinct rates; and SALM post-insertion `multimodal_tokens` for every SALM variant.
- Support dynamic batches and packed SALM sequences through observed length tensors and SALM's exact post-insertion multimodal-token counter.
- Report exact rank-local example totals and rates, plus effective `mean_batch_size` over optimizer steps so gradient accumulation is included.
- Report every available work unit both per second and per optimizer step; intentionally emit no `_per_example` metrics.
- Align normal reporting windows to optimizer-step boundaries; lifecycle-forced partial windows retain totals and rates but omit misleading per-step means.
- Never report global batch size, micro batch size, or a static sequence length; modalities remain separate units.
- Avoid per-batch host scalar materialization, distributed collectives, and Megatron dependencies. GPU totals and timing are resolved asynchronously with at most one bounded pending window.
- Skip models and measurements whose semantics cannot be determined safely.
- Contain provider, policy, timing, and exporter failures so telemetry cannot interrupt training.
- Add public documentation for opt-in setup, reporting cadence, lifecycle events, and model-specific metrics.

# Usage

Set `NEMO_ONE_LOGGER_ENABLED=true` before starting training. Entry points that call `exp_manager()` then attach the integration automatically, and rank zero exports telemetry.

Throughput uses a minimum interval of 100 training batches by default, or a larger `trainer.log_every_n_steps` value. Set `NEMO_ONE_LOGGER_THROUGHPUT_INTERVAL` to a positive batch count to override that minimum. With gradient accumulation, normal windows close at the next optimizer-step boundary.

# Validation

- `pre-commit run --from-ref origin/main --to-ref HEAD`
- `pytest -q --cpu tests/lightning/test_speech_throughput.py tests/lightning/test_one_logger_callback.py tests/utils/test_training_stats_callback.py` — 71 passed
- Fresh-process imports with all `nv_one_logger` modules blocked — passed for both disabled and explicitly enabled modes
- Focused SALM and SALMAutomodel training-stat regressions — 2 passed
- `pylint --rcfile .pylintrc.other ...` and `pylint --rcfile .pylintrc.speech ...` — 10.00/10
- `flake8 --config .flake8.other ...` and `flake8 --config .flake8.speech ...`
- `uv run make -C docs html` — succeeded; only pre-existing dependency and toctree warnings
- `git diff --check origin/main`
- Remote and local DCO verification for every branch commit
- Two throughput-specific architecture/performance review and iteration rounds; no blockers remain

# GitHub Actions CI

Trusted PRs run automatically through copy-pr-bot.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Read and followed the contributor guidelines
- [x] Added focused tests for the changed behavior
- [x] Existing OneLogger dependencies and import behavior were verified

**PR Type**:

- [x] New Feature
- [x] Bugfix
- [x] Documentation

# Additional Information

No public API removals or incompatible configuration changes.

### CODE REVIEW

File: nemo/lightning/one_logger_callback.py [hunk 1]
```diff
@@ -162,42 +223,57 @@
...
+                    break
+
+    micro_batch_size = micro_batch_size or 1
+
+    data_parallel_size = _get_world_size()
+    get_data_parallel_size = getattr(datamodule, '_get_world_size', None)
+    if isinstance(data_cfg, Mapping) and callable(get_data_parallel_size):
+        try:
+            data_parallel_size = int(get_data_parallel_size())
+        except (AttributeError, KeyError, RuntimeError, TypeError, ValueError):
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/lightning/one_logger_callback.py]: Fixed in 6090d9a8f2: the exception handler now explicitly restores the environment-derived world-size fallback.

File: tests/lightning/test_one_logger_callback.py [hunk 2]
```diff
@@ -278,6 +320,154 @@
...
+
+    @pytest.mark.unit
+    def test_callback_context_always_emits_end(self):
+        group = MagicMock()
+        with patch('nemo.lightning.callback_group.CallbackGroup.get_instance', return_value=group):
+            with pytest.raises(RuntimeError, match="boom"):
+                with callback_context('on_model_init_start', 'on_model_init_end'):
+                    raise RuntimeError("boom")
+
+        group.on_model_init_start.assert_called_once_with()
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply tests/lightning/test_one_logger_callback.py:483]: Fixed in 6090d9a8f2: the exception-path test now catches and asserts the expected error explicitly, leaving the callback assertions visibly reachable.

File: tests/lightning/test_one_logger_callback.py [hunk 3]
```diff
@@ -278,6 +320,154 @@
...
+        group._callbacks = [callback]
+        trainer = SimpleNamespace(callbacks=[])
+
+        group.attach_to_trainer(trainer)
+
+        assert trainer.callbacks == []
+
+    @pytest.mark.unit
+    def test_modelpt_subclass_init_emits_one_paired_span(self):
+        class ParentModel(ModelPT):
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply tests/lightning/test_one_logger_callback.py:432]: Fixed in 6090d9a8f2: the test model now calls `ModelPT.__init__` with an empty config.

File: nemo/core/classes/modelPT.py [hunk 4]
```diff
@@ -145,55 +148,65 @@
...
+        with dataloader_context:
+            if self._cfg is not None and not self._is_model_being_restored():
+                # Setup data loaders now (default) or defer setup to `self.setup()`
+                # if `defer_setup` is set in the config of the corresponding dataloader.
+                if (
+                    'train_ds' in self._cfg
+                    and self._cfg.train_ds is not None
+                    and not self._cfg.train_ds.get('defer_setup', False)
+                ):
+                    self.setup_training_data(self._cfg.train_ds)
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/core/classes/modelPT.py:173]: This call intentionally predates the PR: ModelPT eager dataloader setup invokes the model-specific override during construction. This change only wraps that existing call in an exception-safe telemetry context and does not introduce or alter the virtual dispatch.

File: nemo/core/classes/modelPT.py [hunk 5]
```diff
@@ -145,55 +148,65 @@
...
+                    and not self._cfg.validation_ds.get('defer_setup', False)
+                ):
+                    self.setup_multiple_validation_data(val_data_config=cfg.validation_ds)
+
+                if (
+                    'test_ds' in self._cfg
+                    and self._cfg.test_ds is not None
+                    and not self._cfg.test_ds.get('defer_setup', False)
+                ):
+                    self.setup_multiple_test_data(test_data_config=cfg.test_ds)
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/core/classes/modelPT.py:187]: This call intentionally predates the PR: ModelPT eager test-dataloader setup invokes the model-specific override during construction. The PR only adds the surrounding telemetry context; dispatch and execution order are unchanged.

File: tests/lightning/test_one_logger_callback.py [hunk 6]
```diff
@@ -1,498 +1,437 @@
...
-        trainer.lightning_module = pl_module
+    def test_callback_context_always_emits_end(self):
+        group = MagicMock()
+        with patch('nemo.lightning.callback_group.CallbackGroup.get_instance', return_value=group):
+            with pytest.raises(RuntimeError, match="boom"):
+                with callback_context('on_model_init_start', 'on_model_init_end'):
+                    raise RuntimeError("boom")
 
-        config = get_nemo_v1_callback_config(trainer)
+        group.on_model_init_start.assert_called_once_with()
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply tests/lightning/test_one_logger_callback.py:742]: False positive: `pytest.raises` consumes the expected exception when its context exits, after which both assertions are reachable and execute. The focused test suite passes this test.

File: nemo/core/classes/modelPT.py [hunk 7]
```diff
@@ -145,55 +148,65 @@
...
+                    and not self._cfg.train_ds.get('defer_setup', False)
+                ):
+                    self.setup_training_data(self._cfg.train_ds)
+
+                if (
+                    'validation_ds' in self._cfg
+                    and self._cfg.validation_ds is not None
+                    and not self._cfg.validation_ds.get('defer_setup', False)
+                ):
+                    self.setup_multiple_validation_data(val_data_config=cfg.validation_ds)
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/core/classes/modelPT.py:180]: This call intentionally predates the PR: ModelPT eager validation-dataloader setup invokes the model-specific override during construction. The PR only adds the surrounding telemetry context; dispatch and execution order are unchanged.

File: nemo/collections/tts/models/easy_magpietts_cfg_distillation.py [hunk 8]
```diff
@@ -57,7 +57,7 @@
     transcribe_with_whisper,
     transcribe_with_whisper_from_filepaths,
 )
-from nemo.lightning.callback_group import CallbackGroup
+from nemo.lightning.callback_group import with_callback_context
```
Comment (blisc) [inline nemo/collections/tts/models/easy_magpietts_cfg_distillation.py:60]: Do you know what these changes do?

Comment (pzelasko) [reply nemo/collections/tts/models/easy_magpietts_cfg_distillation.py:60]: It adds a PTL callback that collects metrics and events+timing to report to OneLogger on relevant events (model/dataloader/optimizer init, training step begin/end, etc). I agree these names are way too vague, will improve.

**Review by blisc — APPROVED** [review]
